### PR TITLE
Fix label list order for h-label classification

### PR DESCRIPTION
**Pull request by GalyaZalesskaya**
### Summary

This PR fixes the issue with wrong saliency maps for h-label classification.

It is caused by the fact that the order for labels in `label_schema` doesn't correspond with the model outputs order of labels for h-label cls. So then saliency map for the predicted class is added, it uses the wrong index that results in adding the saliency map for the wrong class. 

Not to interfere with the logic of `label_schema`, I added the function `get_hierarchical_label_list` to return the list or labels in order that is used in model output. I use it to return the saliency map for the right predicted class.

Please, make your comment if you suggest moving the function to the different `*_utils.py` file.

Issue: CVS-117964

Saliency map for "young men" class before and after changes:
image

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/api/utils/dataset_utils.py [hunk 1]
```diff
@@ -272,3 +272,15 @@ def non_linear_normalization(saliency_map: np.ndarray) -> np.ndarray:
     saliency_map = 255.0 / (max_soft_score + 1e-12) * saliency_map
 
     return np.uint8(np.floor(saliency_map))
+
+
+def get_hierarchical_label_list(hierarchical_info, labels):
```
Comment (sovrasov) [inline src/otx/api/utils/dataset_utils.py]: Classification utils are located in src/otx/algorithms/classification/utils/cls_utils.py

File: src/otx/api/utils/dataset_utils.py [hunk 2]
```diff
@@ -272,3 +272,15 @@ def non_linear_normalization(saliency_map: np.ndarray) -> np.ndarray:
...
+    """Return hierarchical labels list which is adjusted to model outputs classes."""
+    hierarchical_labels = []
+    for head_idx in range(hierarchical_info["num_multiclass_heads"]):
+        logits_begin, logits_end = hierarchical_info["head_idx_to_logits_range"][str(head_idx)]
+        for logit in range(0, logits_end - logits_begin):
+            label_str = hierarchical_info["all_groups"][head_idx][logit]
+            otx_label = next(x for x in labels if x.name == label_str)
+            hierarchical_labels.append(otx_label)
```
Comment (sovrasov) [inline src/otx/api/utils/dataset_utils.py]: Multilabel heads should also be considered here

File: src/otx/api/utils/dataset_utils.py [hunk 3]
```diff
@@ -272,3 +272,15 @@ def non_linear_normalization(saliency_map: np.ndarray) -> np.ndarray:
...
+
+
+def get_hierarchical_label_list(hierarchical_info, labels):
+    """Return hierarchical labels list which is adjusted to model outputs classes."""
+    hierarchical_labels = []
+    for head_idx in range(hierarchical_info["num_multiclass_heads"]):
+        logits_begin, logits_end = hierarchical_info["head_idx_to_logits_range"][str(head_idx)]
+        for logit in range(0, logits_end - logits_begin):
+            label_str = hierarchical_info["all_groups"][head_idx][logit]
+            otx_label = next(x for x in labels if x.name == label_str)
```
Comment (sovrasov) [inline src/otx/api/utils/dataset_utils.py]: Label 2 idx can be applied here, like it's done in annotation converter:
```
label_idx = hierarchical_info["label_to_idx"][label_str]
label = labels[label_idx ]
```

File: src/otx/api/utils/dataset_utils.py (hunk 1 above)
Comment (GalyaZalesskaya) [reply src/otx/api/utils/dataset_utils.py]: Thank you, done

File: src/otx/api/utils/dataset_utils.py (hunk 2 above)
Comment (GalyaZalesskaya) [reply src/otx/api/utils/dataset_utils.py]: Done

File: src/otx/api/utils/dataset_utils.py (hunk 3 above)
Comment (GalyaZalesskaya) [reply src/otx/api/utils/dataset_utils.py]: Great point for optimizing this, thank you!

**Review by sovrasov — DISMISSED** [review]

**Review by sovrasov — APPROVED** [review]

**Review by goodsong81 — APPROVED** [review]
Thank you for the fix!
